internal/clientconn/listener.go

```diff
@@ -69,23 +74,75 @@ func NewListener(opts *NewListenerOpts) *Listener {
 func (l *Listener) Run(ctx context.Context) error {
 	logger := l.opts.Logger.Named("listener")
 
-	var err error
-	if l.listener, err = net.Listen("tcp", l.opts.ListenAddr); err != nil {
-		return lazyerrors.Error(err)
+	useSock := l.opts.ListenSock != ""
+	useTcp := !useSock || l.opts.ListenAddr != defaultListeningIP
+
+	if useTcp {
+		var err error
+		if l.tcpListener, err = net.Listen("tcp", l.opts.ListenAddr); err != nil {
+			return lazyerrors.Error(err)
+		}
+	}
+
+	if useSock {
+		var err error
+		if l.sockListener, err = net.Listen("unix", l.opts.ListenSock); err != nil {
+			return lazyerrors.Error(err)
+		}
 	}
 
 	close(l.listening)
-	logger.Sugar().Infof("Listening on %s ...", l.Addr())
+
+	if useTcp {
+		logger.Sugar().Infof("Listening on %s ...", l.Addr())
```

Comment:
I think we could move that logging to `if`s above

Reply:
If the port is zero (`--listen-addr=:0`), a random port will be allocated by OS, but we will know the number only after the listener starts

Reply:
I guess we could rename `listening` to `tcpListenerReady` and use it only in `Addr()`

Reply:
> If the port is zero (--listen-addr=:0), a random port will be allocated by OS, but we will know the number only after the listener starts

But why you restrict the value to a default in case of embedded usage?